packages/ampd-sdk/src/grpc/connection_manager.rs

```diff
@@ -0,0 +1,216 @@
+#[derive(Debug)]
+pub struct ConnectionManager {
+    url: Url,
+    connection_state: watch::Sender<ConnectionState>,
+    message_receiver: mpsc::Receiver<ClientMessage>,
+    retry_policy: RetryPolicy,
+}
+
+impl ConnectionManager {
+    pub fn new(url: &str) -> Result<(Self, ConnectionHandle), Error> {
+        let url = Url::new_sensitive(url).change_context(AppError::InvalidUrl.into())?;
+
+        let (state_sender, state_receiver) = watch::channel(ConnectionState::Disconnected);
+        let (msg_sender, msg_receiver) = mpsc::channel(100);
+
+        let handle = ConnectionHandle {
+            connection_receiver: state_receiver,
+            message_sender: msg_sender,
+        };
+
+        let manager = Self {
+            url,
+            connection_state: state_sender,
+            message_receiver: msg_receiver,
+            retry_policy: DEFAULT_RETRY_POLICY,
+        };
+
+        Ok((manager, handle))
+    }
+
+    pub async fn run(mut self) -> Result<(), Error> {
+        info!("starting the connection manager");
+        let _ = self.connect().await;
+
+        while let Some(message) = self.message_receiver.recv().await {
+            match message {
+                ClientMessage::ConnectionFailed(details) => {
+                    warn!("client reported connection failure: {}", details);
+                    self.handle_connection_failure().await;
+                }
+            }
+        }
+
+        Ok(())
+    }
+
+    async fn connect(&mut self) -> Result<(), Error> {
+        let endpoint: transport::Endpoint = self.url.as_str().parse().into_report()?;
+        let endpoint = endpoint
+            .connect_timeout(DEFAULT_INITIAL_TIMEOUT)
+            .timeout(DEFAULT_RPC_TIMEOUT)
+            .keep_alive_timeout(KEEPALIVE_TIMEOUT)
+            .keep_alive_while_idle(KEEPALIVE_WHILE_IDLE)
+            .http2_keep_alive_interval(KEEPALIVE_TIME);
+
+        match endpoint.connect().await {
+            Ok(channel) => {
+                info!("successfully connected to ampd gRPC server");
+                let _ = self
+                    .connection_state
+                    .send(ConnectionState::Connected(channel));
+                Ok(())
+            }
+            Err(status) => {
+                let _ = self.connection_state.send(ConnectionState::Disconnected);
+                warn!(err = ?status, "reconnecting to the ampd gRPC server failed");
+                Err(status).into_report()
+            }
+        }
+    }
+
+    async fn handle_connection_failure(&mut self) {
+        let _ = self.connection_state.send(ConnectionState::Reconnecting);
+
+        let max_attempts = self.retry_policy.max_attempts();
+        let mut attempts = 0u64;
+
+        while attempts < max_attempts {
+            attempts = attempts.saturating_add(1);
+
+            if let Some(delay) = self.retry_policy.delay() {
+                tokio::time::sleep(delay).await;
+            }
```

Comment:
why do you insist on reimplementing the abilities of the retry policy? 😆 
just use `with_retry`